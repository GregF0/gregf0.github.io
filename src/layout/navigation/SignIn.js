import React, { useState } from 'react'
import { useDispatch } from 'react-redux'
import styled from 'styled-components'
import Button from '@material-ui/core/Button'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import FaceIcon from '@material-ui/icons/Face'
import ListItemIcon from '@material-ui/core/ListItemIcon'
import ListItemText from '@material-ui/core/ListItemText'
import SvgIcon from '@material-ui/core/SvgIcon'
import { GoogleIcon } from './GoogleIcon'
import { linkAccount } from '../../data/auth/actions'
import { googleAuthProvider } from '../../firebase/firebase'
import { setProgressVisibility } from '../../data/progress/actions'

export const SignIn = () => {
  const [anchorEl, setAnchorEl] = useState(null)
  const dispatch = useDispatch()

  const googleAuth = async () => {
    dispatch(setProgressVisibility(true))
    handleClose()
    await dispatch(linkAccount(googleAuthProvider))
    dispatch(setProgressVisibility(false))
  }

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <>
      <TriggerButton
        aria-controls="sign-in-menu"
        aria-haspopup="true"
        onClick={openMenu}
        startIcon={<FaceIcon />}
      >
        Sign In
      </TriggerButton>
      <Menu
        id="sign-in-menu"
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={handleClose}
      >
        <MenuItem onClick={googleAuth}>
          <ListItemIcon>
            <SvgIcon>
              <GoogleIcon />
            </SvgIcon>
          </ListItemIcon>
          <ListItemText primary="Sign in via Google" />
        </MenuItem>
      </Menu>
    </>
  )
}

export const TriggerButton = styled(Button)`
  color: #fff;
`
