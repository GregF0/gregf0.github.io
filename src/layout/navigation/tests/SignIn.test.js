import React from 'react'
import * as redux from 'react-redux'
import { createShallow } from '@material-ui/core/test-utils'
import Menu from '@material-ui/core/Menu'
import { SignIn, TriggerButton } from '../SignIn'

describe('<SignIn />', () => {
  const shallow = createShallow()
  const createWrapper = () => {
    return shallow(<SignIn />)
  }

  const dispatchMocked = jest.fn()
  jest.spyOn(redux, 'useDispatch').mockImplementation(() => dispatchMocked)

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render <SignIn /> correctly', () => {
    expect(createWrapper()).toMatchSnapshot()
  })

  test('menu should be closed initially', () => {
    expect(createWrapper().find(Menu).props().open).toBe(false)
  })

  test('should open menu on avatar click', () => {
    const wrapper = createWrapper()

    wrapper.find(TriggerButton).simulate('click', { currentTarget: {} })
    expect(wrapper.find(Menu).props().open).toBe(true)
  })
})
