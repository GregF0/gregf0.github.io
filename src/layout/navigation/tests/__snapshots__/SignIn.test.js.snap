// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn /> should render <SignIn /> correctly 1`] = `
<Fragment>
  <Styled(WithStyles(ForwardRef(Button)))
    aria-controls="sign-in-menu"
    aria-haspopup="true"
    onClick={[Function]}
    startIcon={<UNDEFINED />}
  >
    Sign In
  </Styled(WithStyles(ForwardRef(Button)))>
  <WithStyles(ForwardRef(Menu))
    anchorEl={null}
    id="sign-in-menu"
    keepMounted={true}
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(MenuItem))
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(SvgIcon))>
          <GoogleIcon />
        </WithStyles(ForwardRef(SvgIcon))>
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Sign in via Google"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Menu))>
</Fragment>
`;
